lib/src/util/condition_scope_visitor.dart

```diff
@@ -0,0 +1,307 @@
+      environment.add(e);
+    }
+  }
+
+  void addAll(Iterable<_ExpressionBox> expressions) {
+    environment.addAll(expressions);
+  }
+
+  Iterable<Expression> getExpressions(Iterable<Element> elements, bool value) {
+    final expressions = <Expression>[];
+    _recursiveGetExpressions(this, expressions, elements, value);
+    return expressions;
+  }
+
+  Iterable<_ExpressionBox> getUndefinedExpressions() =>
+      environment.where((e) => e is _UndefinedExpression);
+
+  static void _recursiveGetExpressions(ConditionScope conditionScope,
+      List<Expression> expressions, Iterable<Element> elements, bool value) {
+    if (conditionScope == null) {
+      return;
+    }
+    for (final element in conditionScope.environment.reversed) {
+      if (element.haveToStop(elements)) {
+        return;
+      }
+      if (element is _ConditionExpression && element.value == value) {
+        expressions.add(element.expression);
+      }
+    }
+    _recursiveGetExpressions(
+        conditionScope.outer, expressions, elements, value);
+  }
+}
+
+abstract class ConditionScopeVisitor extends UnifyingAstVisitor {
+  ConditionScope outerScope;
+
+  Iterable<Expression> getFalseExpressions(Iterable<Element> elements) =>
+      _getExpressions(elements, value: false);
+
+  Iterable<Expression> getTrueExpressions(Iterable<Element> elements) =>
+      _getExpressions(elements);
+
+  @override
+  visitAssignmentExpression(AssignmentExpression node) {
+    _addElementToEnvironment(new _UndefinedExpression(_getLeftElement(node)));
+    node.visitChildren(this);
+  }
+
+  @override
+  visitBlockFunctionBody(BlockFunctionBody node) {
+    _addScope();
+    _addElementToEnvironment(new _UndefinedAllExpression());
+    node.visitChildren(this);
+    _removeLastScope();
+  }
+
+  void visitCondition(Expression node);
+
+  @override
+  visitDoStatement(DoStatement node) {
+    _addScope();
+    visitCondition(node.condition);
+    node.visitChildren(this);
+    _propagateUndefinedExpressions(_removeLastScope());
+  }
+
+  @override
+  visitEmptyFunctionBody(EmptyFunctionBody node) {
+    _addScope();
+    _addElementToEnvironment(new _UndefinedAllExpression());
+    node.visitChildren(this);
+    _removeLastScope();
+  }
+
+  @override
+  visitExpressionFunctionBody(ExpressionFunctionBody node) {
+    _addScope();
+    _addElementToEnvironment(new _UndefinedAllExpression());
+    node.visitChildren(this);
+    _removeLastScope();
+  }
+
+  @override
+  visitForStatement(ForStatement node) {
+    _addScope();
+    _addTrueCondition(node.condition);
+    node.variables?.accept(this);
+    node.initialization?.accept(this);
+    visitCondition(node.condition);
+    node.condition?.accept(this);
+    _addTrueCondition(node.condition);
+    node.updaters.accept(this);
+    node.body?.accept(this);
+    _propagateUndefinedExpressions(_removeLastScope());
+    bool hasBreak = DartTypeUtilities
+        .traverseNodesInDFS(node, excludeCriteria: _isLoopStatement)
```

Comment:
My understanding is that `hasBreak` should be `true` (in both the `for` and `while` cases) if there is any way for control to reach the statement following the loop other than for the condition to evaluate to `false`. But if you don't look inside the nested `while` loop (in the example above) then you won't find the `break` statement that applies to the `for` loop and `hasBreak` will be `false` despite the fact that the condition doesn't have to be `false`. Here's a more concrete example:
```
String s = '...';
loop: for (; s != null; s += '.') {
  while (true != false) {
    break loop;
  }
}
// I think there will be a lint on the following line,
// even though the test is reasonable.
if (s != null) {
  print(s);
}
```